packages/rich-list-core/src/RichListCore.ts

```diff
@@ -0,0 +1,226 @@
+      throw new Error('Invalid token id')
+    }
+
+    if (!(await this._listTokenIds()).includes(Number(tokenId))) {
+      throw new Error('Invalid token id')
+    }
+
+    return (await this.existingRichList.list({
+      partition: `${tokenId}`,
+      limit: this.richListLength,
+      order: 'DESC'
+    })).map(s => s.data)
+  }
+
+  /**
+   * Updated rich list with latest balance
+   * by sorting the existing rich list together with `queuedAddressLimit` number of recently active addresses
+   *
+   * @param queuedAddressLimit [5000]
+   */
+  async calculateNext (queuedAddressLimit = 5000): Promise<void> {
+    const tokens = await this._listTokenIds()
+    let updatedBalances = await this._getActiveAddressBalances(tokens, queuedAddressLimit)
+
+    while (Object.keys(updatedBalances).length > 0) {
+      for (const address of Object.keys(updatedBalances)) {
+        const accountAmount = updatedBalances[address]
+        for (const tokenId of tokens) {
+          const balance = accountAmount[tokenId].toNumber()
+          await this.existingRichList.put({
+            partition: `${tokenId}`,
+            sort: balance,
+            id: `${address}-${tokenId}`,
+            data: {
+              address: address,
+              amount: balance
+            }
+          })
+        }
+      }
+
+      updatedBalances = await this._getActiveAddressBalances(tokens, queuedAddressLimit)
+    }
+  }
+
+  private async _getActiveAddressBalances (tokens: number[], queuedAddressLimit: number): Promise<ActiveAddressAccountAmount> {
+    const queue = await this._addressQueue()
+    const addresses = await queue.receive(queuedAddressLimit)
+
+    const balances: { [key: string]: AccountAmount } = {}
+    for (const a of addresses) {
+      const nonZeroBalances = await this.apiClient.account.getTokenBalances(
+        { limit: Number.MAX_SAFE_INTEGER },
+        true
+      ) as any as AccountAmount
+      balances[a] = this._appendZeroBalances(nonZeroBalances, tokens)
+    }
+    return balances
+  }
+
+  private _appendZeroBalances (tokenBalances: AccountAmount, tokens: number[]): AccountAmount {
+    const result: AccountAmount = {}
+    for (const t of tokens) {
+      result[t] = tokenBalances[t] ?? new BigNumber(0)
+    }
+    return result
+  }
+
+  private async _getCrawledTip (): Promise<Schema<CrawledBlock> | undefined> {
+    const [lastBlock] = await this.crawledBlockHashes.list({
+      partition: 'NONE',
+      order: 'DESC',
+      limit: 1
+    })
+    return lastBlock
+  }
+
+  private async _invalidateDroppedOutRichList (tokenId: string, height: number): Promise<void> {
+    const lastBlockRichList = await this._getDroppedOutRichList(tokenId, height)
+    await Promise.all(
+      lastBlockRichList.map(async rl => await this.droppedFromRichList.delete(rl.id))
+    )
+  }
+
+  private async _getDroppedOutRichList (tokenId: string, height: number): Promise<Array<Schema<string>>> {
+    return await this.droppedFromRichList.list({
```

Comment:
Should be equivalent to
```suggestion
    return this.droppedFromRichList.list({
```

Reply:
We are currently employing explicit `await` in Jellyfish so the above version is expected.